File: lesson-520-developing-with-rdds/solutions/python/solution-02-parallelism/solution.py
from operator import add
from datetime import datetime

# get the word with bigger count
def getMax(r, c):
  if (r[1] > c[1]):
    return r
  else:
    return c

srcDir = <set srcDir to the absolute path of the source data directory location>
file = srcDir + "/sherlock-holmes.txt"

# read text and cache
text = sc.textFile(file).persist()

minCores = 2
numCores = 8

for prt in range(minCores, numCores+1, 2):

    rept = text.repartition(prt)

    dt1 = datetime.now()
    words = rept.flatMap(lambda w : w.split(' '))
    dict = words.map(lambda w : (w, 1))
    counts = dict.reduceByKey(add)
    max = counts.reduce(lambda r, c: getMax(r, c))
    dt2 = datetime.now()
    print(str(max) + " partition count " + str(prt) + " time " + str((dt2-dt1).microseconds/1000))

File: lesson-520-developing-with-rdds/solutions/scala/solution-02-parallelism/solution.scala
// NOTE: set the value srcDir to a string containing the absolute path of
// the location of the source data directory, lesson-5xx-resources!

val minCores = 2 // if your system has 2+ cores
val nCores = 8 // depends on your system
val partitions = 2 to nCores by 2
val files = Array("byu-corpus.txt"/*, "sherlock-holmes.txt"*/);
val delimiters = Array(' ',',','.',':',';','?','!','-','@','#','(',')');
files foreach( file => {
  val original = sc.textFile(srcDir + "/" + file);
  println("File: " + file);
  partitions foreach(n => {
    println("Partitions: " + n);
    val partitioned = original.repartition(n);
    val start = System.currentTimeMillis;
    val result = partitioned.flatMap(_.split(delimiters)).filter(_.trim.length > 0).map(x => (x, 1)).reduceByKey(_ + _).sortBy(_._2, false).first;
    val time = System.currentTimeMillis - start;
    println("Took: " + time + " ms");
  });
});
